subsys/net/lib/conn_mgr/conn_mgr.c

```diff
@@ -0,0 +1,197 @@
+}
+#else
+#define conn_mgr_ipv6_status(...) NET_CONN_MGR_STATE_CONNECTED
+#endif /* CONFIG_NET_IPV6 */
+
+#if defined(CONFIG_NET_IPV4)
+static enum net_conn_mgr_state conn_mgr_ipv4_status(int iface_idx)
+{
+	if ((iface_states[iface_idx] & CONN_MGR_IPV4_STATUS_MASK) ==
+	    CONN_MGR_IPV4_STATUS_MASK) {
+		NET_DBG("IPv4 connected on iface index %u", iface_idx);
+		return NET_CONN_MGR_STATE_CONNECTED;
+	}
+
+	return NET_CONN_MGR_STATE_DISCONNECTED;
+}
+#else
+#define conn_mgr_ipv4_status(...) NET_CONN_MGR_STATE_CONNECTED
+#endif /* CONFIG_NET_IPV4 */
+
+static void conn_mgr_notify_status(int iface_idx)
+{
+	struct net_if *iface = net_if_get_by_index(iface_idx + 1);
+
+	if (iface_states[iface_idx] & NET_STATE_CONNECTED) {
+		NET_DBG("Iface %p connected", iface);
+		net_mgmt_event_notify(NET_EVENT_L4_CONNECTED, iface);
+	} else {
+		NET_DBG("Iface %p disconnected", iface);
+		net_mgmt_event_notify(NET_EVENT_L4_DISCONNECTED, iface);
+	}
+}
+
+static void conn_mgr_act_on_changes(void)
+{
+	int idx;
+
+	for (idx = 0; idx < CONN_MGR_IFACE_MAX; idx++) {
+		enum net_conn_mgr_state state;
+
+		if (!(iface_states[idx] & NET_STATE_CHANGED)) {
+			continue;
+		}
+
+		state = NET_CONN_MGR_STATE_CONNECTED;
+
+		state &= conn_mgr_iface_status(idx);
+		if (state) {
+			enum net_conn_mgr_state ip_state =
+				NET_CONN_MGR_STATE_DISCONNECTED;
+
+			if (IS_ENABLED(CONFIG_NET_IPV6)) {
+				ip_state |= conn_mgr_ipv6_status(idx);
+			}
+
+			if (IS_ENABLED(CONFIG_NET_IPV4)) {
+				state |= conn_mgr_ipv4_status(idx);
+			}
+
+			state &= ip_state;
+		}
+
+		iface_states[idx] &= ~NET_STATE_CHANGED;
+
+		if (state == NET_CONN_MGR_STATE_CONNECTED &&
+		    !(iface_states[idx] & NET_STATE_CONNECTED)) {
+			iface_states[idx] |= NET_STATE_CONNECTED;
+
+			conn_mgr_notify_status(idx);
+		} else if (state != NET_CONN_MGR_STATE_CONNECTED &&
+			   (iface_states[idx] & NET_STATE_CONNECTED)) {
+			iface_states[idx] &= ~NET_STATE_CONNECTED;
+
+			conn_mgr_notify_status(idx);
+		}
+	}
+}
+
+static void conn_mgr_initial_state(struct net_if *iface)
+{
+	int idx = net_if_get_by_iface(iface) - 1;
+
+	if (net_if_is_up(iface)) {
+		NET_DBG("Iface %p UP", iface);
+		iface_states[idx] = NET_STATE_IFACE_UP;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV6)) {
+		if (net_if_ipv6_get_global_addr(NET_ADDR_PREFERRED,
+						&iface)) {
+			NET_DBG("IPv6 addr set");
+			iface_states[idx] |= NET_STATE_IPV6_ADDR_SET |
+				NET_STATE_IPV6_DAD_OK;
+		} else if (net_if_ipv6_get_global_addr(NET_ADDR_TENTATIVE,
+						       &iface)) {
+			iface_states[idx] |= NET_STATE_IPV6_ADDR_SET;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_NET_IPV4)) {
+		if (net_if_ipv4_get_global_addr(iface,
+						NET_ADDR_PREFERRED)) {
+			NET_DBG("IPv4 addr set");
+			iface_states[idx] |= NET_STATE_IPV4_ADDR_SET;
+		}
+
+	}
+
+	iface_states[idx] |= NET_STATE_CHANGED;
+}
+
+static void conn_mgr_init_cb(struct net_if *iface, void *user_data)
+{
+	ARG_UNUSED(user_data);
+
+	conn_mgr_initial_state(iface);
+}
+
+static void conn_mgr(void)
+{
```

Comment:
If we need to have a thread for this, could we piggybag this stuff in net_mgmt thread instead?

Reply:
Running things directly in the event callbacks, at this stage why not. But let's say we go for the connectivity check: that's going to be a totally different story (since it will have to be async), we won't be able to lock the net_mgmt thread for this.

Reply:
Sure, that would not run in net_mgmt thread. I would imagine that any connectivity check would be done from separate thread or user application etc which would have more resources.